### PR TITLE
Windowed normalize optimization

### PR DESCRIPTION
**Pull request by jwittenbach**
Added a new option -- 'window-fast' -- to TimeSeries.normalize()
This is similar to the 'window' option, except that interpolation during the percentile computation is avoided by no using the numpy.percentile() method. Instead, the index within the window that is closest to the desired percentile is computed ahead of time. Then each window is sorted and this index is retrieved. Also updated the warning message for when widow size is set but user is not using one of the two windowed methods.

### CODE REVIEW

**Comment by jwittenbach** [comment]
Just realized that I still need to write a new unit test for this method. Will do this and then push the changes.

File: python/thunder/rdds/timeseries.py [hunk 1]
```diff
@@ -298,6 +297,21 @@ def normalize(self, baseline='percentile', window=None, perc=20):
...
+                left, right = (window/2 - 1, window/2)
+
+            n = len(self.index)
+            ind_perc = round((window-1)*(perc/100)) 
+            
+            def basefunc(x):
+                left_pad = mean(x[:right])
+                right_pad = mean(x[-left:])
+                y = concatenate([ left_pad*ones(left), x, right_pad*ones(right) ])  
+                return asarray([sort(y[i-left:i+right+1])[ind_perc] for i in xrange(left,len(x)+left)])
```
Comment (freeman-lab) [inline python/thunder/rdds/timeseries.py]: Insert space after `left,`

File: python/thunder/rdds/timeseries.py [hunk 2]
```diff
@@ -298,6 +297,21 @@ def normalize(self, baseline='percentile', window=None, perc=20):
             basefunc = lambda x: asarray([percentile(x[max(ix-left, 0):min(ix+right+1, n)], perc)
                                           for ix in arange(0, n)])
 
+        if method == 'window-fast':
+            if window%2:
```
Comment (freeman-lab) [inline python/thunder/rdds/timeseries.py]: Insert white space around `%`

File: python/thunder/rdds/timeseries.py [hunk 3]
```diff
@@ -298,6 +297,21 @@ def normalize(self, baseline='percentile', window=None, perc=20):
...
+                left, right = ((window-1)/2, (window-1)/2)
+            else:
+                left, right = (window/2 - 1, window/2)
+
+            n = len(self.index)
+            ind_perc = round((window-1)*(perc/100)) 
+            
+            def basefunc(x):
+                left_pad = mean(x[:right])
```
Comment (freeman-lab) [inline python/thunder/rdds/timeseries.py]: Remove unscores by replacing with `leftPad` and `rightPad`. See the new style guide (to be added in 0.5.0): https://github.com/freeman-lab/thunder/blob/master/python/doc/style_guide.rst

**Comment by freeman-lab** [comment]
Thanks @jwittenbach ! Code looks good except for some style / PEP convention things. You might want to enable PEP checking in PyCharm if you aren't already, that should catch things like missing spaces. And we need to figure out why the new tests are failing. It looks like it's the assertion on line 106, not yet obvious to me what the problem is.

**Comment by jwittenbach** [comment]
I'm running the test locally and debugging that right now. I will push the new tests as well as the formatting as soon as I have it.

**Comment by freeman-lab** [comment]
LGTM!

**Comment by poolio** [comment]
Sorry for the late comment, but you may also be interested in `scipy.ndimage.filters.percentile_filter`  that uses a speedier Cython implementation. I can run some speed tests and submit a PR if you want.

**Comment by freeman-lab** [comment]
@poolio thanks, awesome, didn't know about that. If it's faster would be great if you wanted to submit a PR with it (probably replacing the one we just added).

**Pull request merged**